File: MK-52_ESP32/src/Receiver_Text.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receivers.hpp"

#define __DEBUG

const char _RT_DigitConversion[] PROGMEM =  "### #<=>!,;:?7410852.963--+^E/*##";
//const char _RT_LetterConversion[] PROGMEM = "### #AHOUBIPVCJQWDKRXELSYFMTZGN##";
const char _RT_LetterConversion[] PROGMEM = "#AIQ#BJRXCKSYDLTZEMU.FNV_GOW HP##";

const char _RT_StandardUnity[] PROGMEM = "1.0E+000";

using namespace MK52_Interpreter;

//
// Inits text entry
//
void Receiver_Text::init( void *components[]) {
    #ifdef __DEBUG
    Serial.println("Init text entry");
    #endif
    _text = (char *)malloc(PROGRAM_LINE_LENGTH+1);
    Receiver::init(components);
}

void Receiver_Text::activate( uint8_t scancode, int8_t parent){
    #ifdef __DEBUG
    Serial.println( "Activating receiver TEXT");
    #endif
    Receiver::activate( scancode, parent);
    _setInputMode( 1);
    _text[0] = 0;
    if(!scancode) return;
    tick( scancode);
}

int Receiver_Text::tick( uint8_t scancode){
    int ln = strlen(_text);
    char c = _convertButton( (_mode==2)? _RT_LetterConversion: _RT_DigitConversion, scancode);
    int return_value = NO_CHANGE; 
    switch( scancode){
        //case 1:
        //case 2:
        //    break;
        case 4: // mode change
            _setInputMode( (_mode == 2)? 3: 2);
            break;
        case 31: // enter
            _mode = 0;
            return_value = _parentReceiver;
            break;
        case 32: // Cx
            if( ln<=0){
                _mode = 0;
                return_value = _parentReceiver;
                break;
            }
            _text[--ln] = 0;
            break;
        default: // any other symbol is added to string
            if( ln<=PROGRAM_LINE_LENGTH){
                _text[ln++] = c;
                _text[ln] = 0;
                break;
            }
            _mode = 0;
            return_value = _parentReceiver;
            break;
    }
    return return_value;
}

char *Receiver_Text::toTrimmedString(){
    char *ptr = _text;
    while( *ptr == ' ') ptr++;
    return ptr;
}

char *Receiver_Text::toScreenString(){
    int ln = strlen( _text);
    if( ln<SCREEN_COLS-1) return _text;
    return _text + ln - SCREEN_COLS + 1;
}

//
// Modes:
// mode 1 - receiver takes text from the parent
// mode 2 - input letters
// mode 3 - input digits
//
void Receiver_Text::_setInputMode( uint8_t m, char *parentMode){
    _mode = m;
    switch( m){
        case 2:
            _lcd->updateStatusFMODE( "ALP");
            return;
        case 3:
            _lcd->updateStatusFMODE( "DIG");
            return;
        case 0: //force deactivation
            _text[0] = 0;                        
        default:
            if( parentMode == NULL)
                _lcd->updateStatusFMODE( "   ");
            else
                _lcd->updateStatusFMODE( parentMode);
            return;
    }
}

char *Receiver_Text::appendText( char *text){
    int8_t ln = strlen(_text);
    char *ptr = _text + ln;
    while( ln<PROGRAM_LINE_LENGTH && *text){
        *ptr++ = *text++;
        ln++;
    }
    *ptr = 0;
    return ptr;
}

char *Receiver_Text::appendText_P( const char *text){
    int8_t ln = strlen(_text);
    char *ptr = _text + ln;
    char c = (char)pgm_read_byte(text++);
    while( ln<PROGRAM_LINE_LENGTH && c){
        *ptr++ = c;
        c = (char)pgm_read_byte(text++);
        ln++;
    }
    *ptr = 0;
    return ptr;
}

File: MK-52_Simulator/Receivers/Receiver_Text.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;

namespace MK52Simulator
{
    //
    // Implements a text data entry editor
    //
    public class Receiver_Text : Receiver
    {   
        private const string _RT_DigitConversion =  "### #<=>!,;:?7410852.963--+^E/*##";
        private const string _RT_LetterConversion = "#AIQ#BJRXCKSYDLTZEMU.FNV_GOW HP##";
        private const int _Off = 0;
        private const int _Text = 1;
        private const int _Number = 2;
        //private const int _K = 3;

        /// <summary>
        /// Constructor
        /// </summary>
        public Receiver_Text(MK52_Host parent)
            : base(parent)
        {
            Moniker = "TEXT";
        }

        public override void activate(string prevReceiver)
        {
            base.activate(prevReceiver);
            _mode = _Text;
            _text = new StringBuilder();
            _text.Append(_parent._m_Program_Memory.getCurrentLine());
            updateDisplay( toString());
        }

        public override byte tick(byte scancode)
        {
            RPN_Functions _rpnf = _parent.getFunctions(); 
            switch( scancode)
            {
                case 0:
                    return 0;
                case 4:
                    _mode++;
                    if (_mode > _Number) _mode = _Text;
                    updateDisplay(toString());
                    return 0;
                case 31:
                    completeEntry(toString());
                    return _rpnf.requestNextReceiver(_return_to);
                case 32:
                    if( _text.Length > 0)
                        _text.Remove(_text.Length - 1, 1);
                    updateDisplay(toString());
                    return 0;
                case 33:
                    _rpnf.requestNextReceiver(_return_to);
                    return 33;
                default: 
                    break;
            }

            char c = (char)0;
            switch (_mode)
            {
                case _Text:
                    c = _convertButton(_RT_LetterConversion, scancode);
                    break;
                case _Number:
                    c = _convertButton(_RT_DigitConversion, scancode);
                    break;
                default:
                    break;
            }
            if (c == 0) return 0;
            if (_text.Length >= RPN_Functions.PROGRAM_LINE_LENGTH) return 0;
            _text.Append(c);
            updateDisplay(toString());
            return 0;
        }

        protected virtual void completeEntry(string value)
        {
            _parent._m_Program_Memory.replaceLine(value);
            _parent._m_Program_Memory.incrementCounter();
        }

        protected virtual void updateDisplay(string value)
        {
            LCD_Manager lm = _parent.getLCD();
            StringBuilder sb = new StringBuilder();
            sb.Append(_parent._m_Program_Memory.toCounterString());
            string tmp = toString();
            int ln = tmp.Length - LCD_Manager.SCREEN_COLS + 7;
            if (ln <= 0) sb.Append(tmp);
            else sb.Append(tmp.Substring(ln));
            lm.updateTerminalLine(10, sb.ToString());
            switch (_mode)
            {
                case _Text:
                    lm.updateStatusFMODE("TXT");
                    break;
                case _Number:
                    lm.updateStatusFMODE("NUM");
                    break;
                default:
                    lm.updateStatusFMODE("   ");
                    break;
            }
            lm.requestUpdate();
        }
    }
}
